CheckBox,  Switch: Use `ParameterState`

igotinfected opened the pull request:
## Description
Fixes: #9158 

Makes use of the `ParameterState` to handle value changes.

Other changes:

- `checked` and `intermediate` on the `MudCheckBox` & `MudSwitch` `input`s are now handled correctly
- `MudCheckBox`: `aria-checked` now has `mixed` as value when in `indeterminate` state (i.e. `Value` = `null`)

TODO:

- introduces changes to the `ParameterState` framework that will need to be discussed, check comments
- if `ParameterState` changes are retained they will need to be covered by unit tests

## How Has This Been Tested?

unit tests + visual

## Type of Changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

## Checklist
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

igotinfected committed:
ComponentBase, Field, DateRangePicker: fix unreliable field id

igotinfected committed:
Merge branch 'dev' into fix/mudbooleaninput-value-binding-and-fieldchanged-trigger

igotinfected committed:
Merge branch 'dev' into fix/mudbooleaninput-value-binding-and-fieldchanged-trigger

igotinfected committed:
WIP MudBooleanInput parameter state

igotinfected committed:
Merge branch 'dev' into fix/mudbooleaninput-value-binding-and-fieldchanged-trigger

igotinfected commented:
@ScarletKuro need your input on these changes

#### Why the need for `ParameterStateValueChangeTiming`?

In some cases we shouldn't change the state of the component until the consumer has had the chance to intercept the value and block the change, see #9158.

In the case of `MudCheckBox` + `TriState`, by changing the `_valueState` value before the event callback is invoked it messes with the `TriState` logic that occurs in a specific order (`true` -> `false` -> `null`). It would also cause the checkbox to visually change to the new `_valueState` value, then change back to the value defined by the user in the callback (e.g. from unchecked to checked when clicked, then back to unchecked when the user selects `cancel` in the dialog).

#### Why the need for a fallback `EventCallback`?

If we don't have a fallback to keep `Value` and `_valueState` in sync when the user doesn't provide an `EventCallback`, `Value` will never be updated. A lot of `bUnit` tests that would today be written for example using `Context.RenderComponent<MudCheckbox<bool?>>(...)` would need to have a wrapper test component with a binding to work as expected. A bunch of documentation examples don't use bindings. Same thing for most if not all other form components.

I tried adding `EditorRequired` to `MudBooleanInput`'s `Value` and `ValueChanged` but [found out](https://github.com/dotnet/razor/issues/10553) the razor compiler has some issues with some binding syntax, otherwise that could have helped finding all of the "wrong" usages :/

---

I think the "force users to provide a callback" version is the one you were looking for, it's the one that makes more sense to me too. The fallback version is less work/breaking but introduces complexity to the `ParameterState` framework that isn't really necessary. What do you think?

ScarletKuro commented:
> @ScarletKuro need your input on these changes

I need to find some free time to wrap my head around this.

There is always a debate on how smart `ParameterState` needs to be. While designing it, I had tons of features and settings in mind, but I decided to keep it as simple as possible to ensure it would be solid, reliable, and simple. This change might make sense, and it’s something we potentially need for the `MudForm` rework. However, I still need to analyze the alternatives and determine if this should be part of `ParameterState`. 

To help with this review, it would be great if you could:
1. Write two simple components, preferably without using `MudBooleanInput`, `MudFormComponent`, or similar abstractions:
   1. The first component should use the standard `ParameterState` behavior.
   2. The second component should use your added `ParameterStateValueChangeTiming`.
2. Create a showcase page of these two components and intercept the values you described.

This will help me visually see the difference, making it easier to debug and experiment. This will help us find an alternative or better design. We need to be careful as `ParameterState` is public, although we could consider making some features internal.

@henon, what do you personally think about this?

igotinfected commented:
Sounds good to me, will work on that over the weekend most likely

ScarletKuro commented:
> If we don't have a fallback to keep `Value` and `_valueState` in sync when the user doesn't provide an `EventCallback`, `Value` will never be updated. A lot of `bUnit` tests that would today be written for example using `Context.RenderComponent<MudCheckbox<bool?>>(...)` would need to have a wrapper test component with a binding to work as expected. A bunch of documentation examples don't use bindings. Same thing for most if not all other form components.

As of now, I don’t understand the need for that as it's an Blazor anti-pattern.

While `_valueState` is always in sync with `Value` when you don’t write in the state, I don’t see the reason why the `Value` needs to be in sync when you write to the state. The rule of `ParameterState` and `Blazor` in general is to not write to `[Parameter]`, it should be treated like a testament in the Bible. Your component should never read `Value` directly and should use `_valueState.Value` instead, regardless of whether it’s in `CssBuilder`/`StyleBuilder`, razor syntax, etc. If you need it in another component, then `GetState` should be used.

This is something we’re trying to eliminate. If there is no `EventCallback`, then the underlying `Value` of the component should not change, because only the **parent** has the right to do so, as explained here by a Blazor developer: https://github.com/dotnet/aspnetcore/discussions/49187#discussioncomment-6372174. In simple terms, if the parent is using two-way binding `@bind-XYZ` or a `XYZChanged`(`EventCallback`), then it’s the parent that receives the new value, re-renders itself, and in turn causes the child component to be re-rendered with the new value. It should never be done by anyone else. If it does happen, then the component is written incorrectly.

Let me know if I misunderstood something.

igotinfected commented:
> Let me know if I misunderstood something.

Nope that's exactly it -- I think I complicated things in my head because of the tests and doc examples I saw without callbacks and just wanted to make sure I don't break anything but I should really just be finding all of the affected parts and validate/fix them ;)

igotinfected committed:
Remove ParameterState fallback event callback

ScarletKuro commented:
Another question: Is the idea of `ParameterStateValueChangeTiming.AfterEventCallbacks` just to fire the callback first and then set the value?

If that's the case, isn't there a mistake in the following code?
https://github.com/MudBlazor/MudBlazor/blob/dde88f547f283e3a6bcd880611d32bc28aa5fd6b/src/MudBlazor/State/ParameterStateInternalOfT.cs#L77-L101

On line 92, when it invokes the handler and returns, it will exit the method and line 95 will never be reached.

igotinfected committed:
Merge branch 'dev' into fix/mudbooleaninput-value-binding-and-fieldchanged-trigger

igotinfected committed:
Simplify AfterEventCallback logic & add example component

igotinfected commented:
https://github.com/igotinfected/MudBlazor/blob/dec4abe1906f1a3d205709cb827dbde67bcf1925/src/MudBlazor/State/ParameterStateInternalOfT.cs#L77C5-L92C6

Simplified the code there a little.

> Is the idea of `ParameterStateValueChangeTiming.AfterEventCallbacks` just to fire the callback first and then set the value?

The goal now is to let Blazor handle the update of the state when an event callback is provided.

`ParameterStateValueChangeTiming == Immediate` -> update state immediately
`ParameterStateValueChangeTiming == AfterEventCallback && eventCallback.HasDelegate is false` -> update state immediately
`ParameterStateValueChangeTiming == AfterEventCallback && eventCallback.HasDelegate` -> let Blazor update parameter as usual and state will be updated automatically via `MudComponentBaseWithState`

---

I added two components showing off the issue (`ParameterStateImmediateComp` & `ParameterStateAfterEventCallbackComp`). The code is simple (and obviously inspired by `MudBooleanInput`):

```csharp
// this component's job is to invert the value
public Task ToggleValueAsync()
    => _valueState.SetValueAsync(!_valueState.Value); // , ParameterStateValueChangeTiming.AfterEventCallback);
```

It's hard to explain the difference visually but here's an attempt:


https://github.com/MudBlazor/MudBlazor/assets/15004223/67f71bf4-1fab-4cc5-a94f-be83a00448fd

Basically, despite having a two-way binding, the binding becomes unsynced when the consumer intercepts the value change and "cancels" it (or in this case keeps it unchanged, not triggering `OnParametersSet`).

What do you think? @ScarletKuro

igotinfected committed:
Remove testing stuff

igotinfected committed:
Add `aria-checked` test

henon commented:
> @henon, what do you personally think about this?

I am not sure that changing ParameterState in this way is a good idea, at least I got a bad gut feeling about it. This problem exists in other toolkits as well, for instance in WPF it is solved by introducing a separate "preview change handler" that can be used by the user to intercept the normal change event before it even happens. So the equivalent would be something like this ...

`<MudSwitch @key="Guid.NewGuid().ToString()" T="bool" @bind-Value="Switch" ValueChanging="@OpenDialogAsync" />`

.. where in the handler you get a `ValueChangingEventArgs` object that allows you to manipulate the value before it is applied internally for instance via `OldValue` and `NewValue` properties. By updating the `NewValue` property you can basically cancel the change by doing a `args.NewValue=args.OldValue;`. 

Just an idea, not sure if it is the better solution.

igotinfected commented:
Yeah I didn't really want to touch the `ParameterState` but it seemed like the most logical solution for me -- I've seen `ValueChanging` before in Blazor, unsure where specifically though. Either way I think we'll run into similar issues in the future so it's worth taking the time to decide on a proper implementation.

ScarletKuro commented:
Thanks @igotinfected. I will check all this over the weekend.

> I am not sure that changing ParameterState in this way is a good idea

My idea would be to simply add this interface:
```csharp
public interface IParameterStateUnsafe<T>
{
    T? Value { get; set; }
    EventCallback<T> ValueChanged { get; }
    IEqualityComparer<T> Comparer { get; }
    Task SetValueAsync(T value);
}
```

And make `ParameterStateInternal` implement it. Then you can extend the functionality:
```csharp
public static class ParameterStateExtensions
{
    public static IParameterStateUnsafe<T> AsUnsafe<T>(this ParameterState<T> parameterState) => (IParameterStateUnsafe<T>)parameterState;

    public static async Task SetValueAfterEventCallbackAsync<T>(this IParameterStateUnsafe<T> parameterState, T value)
    {
        if (!parameterState.Comparer.Equals(parameterState.Value, value))
        {
            var eventCallback = parameterState.ValueChanged;
            if (eventCallback.HasDelegate)
            {
                await eventCallback.InvokeAsync(value);
            }

            parameterState.Value = value;
        }
    }
}
```
I tested and it works. Of course, there is a limit on how much you can extend `IParameterState`, since you can't intercept the existing methods, but I think that would be enough and this functionality wouldn't live directly in the core.

ScarletKuro commented:
ups that ready for review was s misclick xD

igotinfected committed:
Merge remote-tracking branch 'remotes/upstream/dev' into fix/mudbooleaninput-value-binding-and-fieldchanged-trigger

ScarletKuro committed:
Use the newly added ReadValue/WriteValueAsync

ScarletKuro commented:
I guess we really need that `AfterEventCallback` for now.

After reviewing other components that use `MudFormComponent`, there are many quirky moments, and if we want to swiftly transition the component to use `ParameterState` without rewriting a lot of code, this seems to be the way. 

However, unlike this solution, I will go with my earlier idea and make it an "unsafe" extension, so that we can easily drop it if we come up with a better idea later.

I will prepare a PR and the `ParameterState` changes can be removed from this PR.

igotinfected commented:
@ScarletKuro awesome, thank you for taking the time!

igotinfected committed:
Merge branch 'dev' into fix/mudbooleaninput-value-binding-and-fieldchanged-trigger

igotinfected committed:
Merge branch 'dev' into fix/mudbooleaninput-value-binding-and-fieldchanged-trigger

ScarletKuro commented:
There was another issue with `MudDrawer`, where a user wanted to intercept the `EventCallback<bool> OpenChanged`, but unfortunately, that’s not possible. This is the main limitation of `ParameterState` - while you can change the value anywhere, you cannot do so within the `EventCallback` of the value you want to modify. I’m struggling to find a good solution.

I want to avoid introducing multiple options, like the proposed `ParameterStateValueChangeTiming`, because:

- It requires fine-tuning depending on the component, which can quickly complicate things.
- Even if we eliminate the two options of `Immediate` and `AfterEventCallback`, making `AfterEventCallback` the default behavior, it would immediately break all components and tests. Any component that uses `ParameterState` would fail, which means that the concept is not successful.

Simply switching the order of `eventCallback.InvokeAsync` and `_value = value` does not directly resolve the issue either. It only eliminates the flicker between the old and new values, assuming you could intercept it. 

At this point, I don’t see a universal solution that could be done in `ParameterState` without breaking existing logic while also allowing for interception.
If we change the logic to always sync the `[Parameter]` value and the state value in `OnParameterSet`, it would yonk the functionality of components like `MudCollapse`. If `Expanded` is not bound, it would become impossible to expand or collapse (depending on the initial `Expanded` value), as the state would constantly be overridden by the parent aka always `true` or `false`.

If we implemented something like this:
```csharp
public override async Task SetValueAsync(T value)
{
	if (!_comparer.Equals(Value, value))
	{
		var eventCallback = _eventCallbackFunc();
		if (eventCallback.HasDelegate)
		{
			await eventCallback.InvokeAsync(value);
		}
		var newValue = _getParameterValueFunc();
		if (!_comparer.Equals(newValue, value))
		{
			// Parent would get new value of parameter after the eventCallback.InvokeAsync so we intercepting the state with new value
			_value = newValue;
			return;
		}
		_value = value;
	}
}
```
It would allow for interception, but again, it would break other logic.

The only viable solution so far is this concept: [MudBlazor Pull Request #9397](https://github.com/MudBlazor/MudBlazor/pull/9397). However, it requires some cumbersome workarounds, such as:

```razor
<MudDrawer @ref="_ref" Open="@_open" OpenChanged="@OpenChanged" Elevation="1" />

@code {
    private bool _open = false;
    private MudDrawer _ref;

    private async Task OpenChanged(bool open)
    {
		// Intercept
		var openState = _ref.GetState(x => x.Open).AsUnsafe();
		openState.Value = false;
    }
}
```

Unfortunately, this approach is not very user-friendly.

 Sorry for the long post.

igotinfected commented:
@ScarletKuro no worries for the long post, I appreciate the detail! Bit busy with life right now so not much time to spare but I'll try looking into this again once I finish up the other things I'm working on (mainly `MudTabs`).
